src/simulators/statevector/statevector_state.hpp

```diff
@@ -634,6 +655,82 @@ void State<statevec_t>::snapshot_matrix_expval(const Operations::Op &op,
   BaseState::qreg_.revert(false);
 }
 
+template <class statevec_t>
+void State<statevec_t>::snapshot_density_matrix(const Operations::Op &op,
+                                                ExperimentData &data,
+                                                SnapshotDataType type) {
+  cmatrix_t reduced_state;
+
+  // Check if tracing over all qubits
+  if (op.qubits.empty()) {
+    reduced_state = cmatrix_t(1, 1);
+    reduced_state[0] = BaseState::qreg_.norm();
+  } else {
+    reduced_state = density_matrix(op.qubits);
+  }
+
+  // Add density matrix to result data
+  switch (type) {
+    case SnapshotDataType::average:
+      data.add_average_snapshot("density_matrix", op.string_params[0],
+                            BaseState::creg_.memory_hex(), std::move(reduced_state), false);
+      break;
+    case SnapshotDataType::average_var:
+      data.add_average_snapshot("density_matrix", op.string_params[0],
+                            BaseState::creg_.memory_hex(), std::move(reduced_state), true);
+      break;
+    case SnapshotDataType::pershot:
+      data.add_pershot_snapshot("density_matrix", op.string_params[0], std::move(reduced_state));
+      break;
+  }
+}
+
+template <class statevec_t>
+cmatrix_t State<statevec_t>::density_matrix(const reg_t &qubits) {
+  return vec2density(qubits, BaseState::qreg_.data());
+}
+
+#ifdef AER_THRUST_SUPPORTED
+template <>
+cmatrix_t State<QV::QubitVectorThrust<float>>::density_matrix(const reg_t &qubits) {
+  return vec2density(qubits, BaseState::qreg_.vector());
+}
+
+template <>
+cmatrix_t State<QV::QubitVectorThrust<double>>::density_matrix(const reg_t &qubits) {
+  return vec2density(qubits, BaseState::qreg_.vector());
+}
```

Comment:
```c++
template <typename floating_t>
cmatrix_t State<QV::QubitVectorThrust<floating_t>>::density_matrix(const reg_t &qubits) {
  return vec2density(qubits, BaseState::qreg_.vector());
}
```
We can do this because both specializations have the same implementation, so the compiler will derive using `monomorphization` (specialize `double` and `float` if they are used in the code at compile time).

Reply:
I tried this change but it resulted in compiler errors like: 

```
error: nested name specifier 'State<QV::DensityMatrixThrust<floating_t> >::' for declaration does not refer into a class, class template or class template partial specialization
cmatrix_t State<QV::DensityMatrixThrust<floating_t>>::reduced_density_matrix(...
```